lib/active_hash/relation.rb

```diff
@@ -7,24 +7,80 @@ class Relation
     delegate :empty?, :length, :first, :second, :third, :last, to: :records
     delegate :sample, to: :records
 
-    def initialize(klass, all_records, query_hash = nil)
+    attr_reader :conditions, :order_values, :klass, :all_records
+
+    def initialize(klass, all_records, conditions = nil, order_values = nil)
       self.klass = klass
       self.all_records = all_records
-      self.query_hash = query_hash
-      self.records_dirty = false
+      self.conditions = Conditions.wrap(conditions || [])
+      self.order_values = order_values || []
+    end
+
+    def where(conditions_hash = :chain)
+      return WhereChain.new(self) if conditions_hash == :chain
+
+      spawn.where!(conditions_hash)
+    end
+
+    class WhereChain
+      attr_reader :relation
+
+      def initialize(relation)
+        @relation = relation
+      end
+
+      def not(conditions_hash)
+        relation.conditions << Condition.new(conditions_hash).invert!
+        relation
+      end
+    end
+
+    def order(*options)
+      spawn.order!(*options)
+    end
+
+    def reorder(*options)
+      spawn.reorder!(*options)
+    end
+
+    def where!(conditions_hash, inverted = false)
+      self.conditions << Condition.new(conditions_hash)
       self
     end
 
-    def where(query_hash = :chain)
-      return ActiveHash::Base::WhereChain.new(self) if query_hash == :chain
+    def spawn
+      self.class.new(klass, all_records, conditions, order_values)
```

Comment:
> thinking out loud (and not a request): Also feels like we should be able to avoid all this passing around `all_records`.

Passing around `all_records` is to support `invert_where` and [other unscoping methods](https://guides.rubyonrails.org/active_record_querying.html#overriding-conditions).

In theory, it could probably be possible to re-load all the records from the main class when a relation is unscoped, but I have a feeling that it would make the behavior of calling `reload` on a chained relation behave unexpected.

Let me know if this is something I should look into.

Reply:
thanks. if it ain't broke...